NullPointerException when trying to create producer tests from Pact

I'm trying to auto-create producer tests from a Pact contract using `spring-cloud-contract-spec-pact` and the verifier in version 1.2.2.BUILD-SNAPSHOT (but I reproduced the error in 1.2.1 as well).

Each time the gradle task `generateContractTests` fails with the following stacktrace:

```
...
Caused by: java.lang.NullPointerException: Cannot get property 'name' on null object
        at org.springframework.cloud.contract.verifier.builder.MethodBodyBuilder.checkType(MethodBodyBuilder.groovy:464)
        at org.springframework.cloud.contract.verifier.builder.MethodBodyBuilder.methodForTypeCheck(MethodBodyBuilder.groovy:422)
        at org.springframework.cloud.contract.verifier.builder.MethodBodyBuilder$_addJsonResponseBodyCheck_closure3.doCall(MethodBodyBuilder.groovy:365)
        at org.springframework.cloud.contract.verifier.builder.MethodBodyBuilder.addJsonResponseBodyCheck(MethodBodyBuilder.groovy:359)
        at org.springframework.cloud.contract.verifier.builder.MethodBodyBuilder.validateResponseBodyBlock(MethodBodyBuilder.groovy:320)
...
```

Here's the pact contract I was using:

```json
{
  "consumer": {
    "name": "ui"
  },
  "provider": {
    "name": "userservice"
  },
  "interactions": [
    {
      "description": "a request to POST a person",
      "providerState": "provider accepts a new person",
      "request": {
        "method": "POST",
        "path": "/user-service/users",
        "headers": {
          "Content-Type": "application/json"
        },
        "body": {
          "firstName": "Arthur",
          "lastName": "Dent"
        }
      },
      "response": {
        "status": 201,
        "headers": {
          "Content-Type": "application/json;charset=UTF-8"
        },
        "body": {
          "id": 42
        },
        "matchingRules": {
          "$.body": {
            "match": "type"
          }
        }
      }
    }
  ],
  "metadata": {
    "pactSpecification": {
      "version": "2.0.0"
    }
  }
}
```

**Further analysis:**
I traced the error back to `MethodBodyBuilder.classToCheck()` where the statement

```java
switch (elementFromBody.class) {
```

should probably be replaced with 
```java
switch (elementFromBody.getClass()) {
```

If `elementFromBody` is a `Map` (as in my case), `elementFromBody.class` in Groovy apparently resolves to the element with the key `class` within the map (which is null) instead of the class of the Map itself.

If the statement above is replaced, the test is generated successfully and evens runs successfully.

--- a/spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/MethodBodyBuilder.groovy
+++ b/spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/MethodBodyBuilder.groovy
@@ -468,7 +468,7 @@ abstract class MethodBodyBuilder {
 
 	// we want to make the type more generic (e.g. not ArrayList but List)
 	protected Class classToCheck(Object elementFromBody) {
-		switch (elementFromBody.class) {
+		switch (elementFromBody.getClass()) {
 			case List:
 				return List
 			case Set:

--- a/spring-cloud-contract-verifier/src/test/groovy/org/springframework/cloud/contract/verifier/builder/MethodBodyBuilderSpec.groovy
+++ b/spring-cloud-contract-verifier/src/test/groovy/org/springframework/cloud/contract/verifier/builder/MethodBodyBuilderSpec.groovy
@@ -487,4 +487,35 @@ DocumentContext parsedJson = JsonPath.parse(json);
 			"JaxRsClientJUnitMethodBodyBuilder"                  | { Contract dsl -> new JaxRsClientJUnitMethodBodyBuilder(dsl, properties) }
 	}
 
+	@Issue("#509")
+	def "classToCheck() should return class of object"() {
+		given:
+			Contract contractDsl = Contract.make {
+				request {
+					method 'POST'
+					url '/api/users'
+				}
+				response {
+					status 200
+				}
+			}
+			MethodBodyBuilder builder = methodBuilder(contractDsl)
+		when:
+			Map<String, String> map = new LinkedHashMap<>()
+			Integer number = Integer.valueOf(42)
+			List<String> list = new ArrayList<>()
+			Set<String> set = new HashSet<>()
+		then:
+			builder.classToCheck(map) == Map.class
+		and:
+			builder.classToCheck(number) == Integer.class
+		and:
+			builder.classToCheck(list) == List.class
+		and:
+			builder.classToCheck(set) == Set.class
+		where:
+			methodBuilderName                                    | methodBuilder
+			"MockMvcSpockMethodBuilder"                          | { Contract dsl -> new MockMvcSpockMethodRequestProcessingBodyBuilder(dsl, properties) }
+	}
+
 }

